### PR TITLE
Refactor how we handle trait bounds in style_derive

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18239)

### CODE REVIEW

File: components/style/values/computed/length.rs [hunk 1]
```diff
@@ -695,28 +695,14 @@ pub type NonNegativeLengthOrNumber = Either<NonNegativeLength, NonNegativeNumber
 /// See values/specified/length.rs for more details.
 #[allow(missing_docs)]
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-#[derive(Clone, Copy, Debug, PartialEq, ToCss)]
+#[derive(Animate, Clone, ComputeSquaredDistance, Copy, Debug, PartialEq)]
+#[derive(ToAnimatedZero, ToCss)]
 pub enum MozLength {
     LengthOrPercentageOrAuto(LengthOrPercentageOrAuto),
+    #[animation(error)]
```
Comment (emilio) [inline components/style/values/computed/length.rs]: Keep the `FIXME`, maybe?

File: components/style/values/generics/basic_shape.rs [hunk 2]
```diff
@@ -44,11 +44,18 @@ add_impls_for_keyword_enum!(ShapeBox);
...
 #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-#[derive(Clone, Debug, PartialEq, ToComputedValue, ToCss)]
+#[derive(Animate, Clone, Debug, PartialEq, ToComputedValue, ToCss)]
 pub enum ShapeSource<BasicShape, ReferenceBox, Url> {
+    #[animation(error)]
     Url(Url),
-    Shape(BasicShape, Option<ReferenceBox>),
+    Shape(
+        BasicShape,
+        #[animation(equal)]
```
Comment (emilio) [inline components/style/values/generics/basic_shape.rs]: What about naming the attribute `only_if_equal`? Or `error_if_unequal`? Seems like it'd be clearer. And the last would be more consistent with `animation(error)`.

File: components/style/values/computed/length.rs [hunk 3]
```diff
@@ -64,26 +63,16 @@ impl ToComputedValue for specified::Length {
     }
 }
 
-#[derive(Clone, Copy, Debug, PartialEq)]
-#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
 #[allow(missing_docs)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Copy, Debug, PartialEq, ToAnimatedZero)]
 pub struct CalcLengthOrPercentage {
+    #[animation(equal)]
```
Comment (emilio) [inline components/style/values/computed/length.rs]: The semantics of this on `ToAnimatedZero` are quite different to `Animate`... I'm not sure how I feel about this.

File: components/style/values/generics/grid.rs [hunk 4]
```diff
@@ -138,19 +138,25 @@ define_css_keyword_enum!{ TrackKeyword:
...
-    Flex(CSSFloat),
+    Flex(
+        #[compute(clone)]
+        CSSFloat,
+    ),
     /// One of the track-sizing keywords (`auto`, `min-content`, `max-content`)
-    Keyword(TrackKeyword),
+    Keyword(
+        #[compute(clone)]
+        TrackKeyword,
```
Comment (emilio) [inline components/style/values/generics/grid.rs]: I think it'd be easier to just `impl ComputedValueAsSpecified for TrackKeyword` in this case, any reason that cannot be done?

File: components/style/properties/longhand/font.mako.rs [hunk 5]
```diff
@@ -1116,13 +1116,13 @@ ${helpers.single_keyword_system("font-variant-caps",
...
-        use values::animated::{Animate, Procedure, ToAnimatedValue, ToAnimatedZero};
-        use values::distance::{ComputeSquaredDistance, SquaredDistance};
+        use values::animated::{ToAnimatedValue, ToAnimatedZero};
 
         #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-        #[derive(Clone, Copy, Debug, PartialEq, ToCss)]
+        #[derive(Animate, Clone, ComputeSquaredDistance, Copy, Debug, PartialEq, ToCss)]
         pub enum T {
             None,
+            #[animation(error)]
```
Comment (emilio) [inline components/style/properties/longhand/font.mako.rs]: Hmm... shouldn't this be in the `None` variant?

**Review by emilio — COMMENTED** [review]
Overall looks good, got a few comments though.

**Comment by nox** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=6e8f1a86c067586de1a03d025562ef5ead853a11

**Comment by nox** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=46ef23b95d92a02108ca9cd47067a0d171bf00a6

Geckolib build error fixed.

**Review by emilio — APPROVED** [review]
Thanks for the documentation! Maybe an example of the generated code would be nice, but that can also be a followup if somebody is confused about it.

r=me

**Pull request merged**